x/app-router/module_ibc.go

```diff
@@ -0,0 +1,364 @@
+func (im IBCModule) OnChanOpenTry(
+	ctx sdk.Context,
+	order channeltypes.Order,
+	connectionHops []string,
+	portID,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	counterpartyVersion string,
+) (string, error) {
+	// doCustomLogic()
+	// core/04-channel/types contains a helper function to split middleware and underlying app version
+	_, cpAppVersion := channeltypes.SplitChannelVersion(counterpartyVersion)
+
+	// call the underlying applications OnChanOpenTry callback
+	version, err := im.app.OnChanOpenTry(
+		ctx,
+		order,
+		connectionHops,
+		portID,
+		channelID,
+		chanCap,
+		counterparty,
+		cpAppVersion, // note we only pass counterparty app version here
+	)
+	if err != nil {
+		return "", err
+	}
+	ctx.Logger().Info(fmt.Sprintf("IBC Chan Open Version %s: ", version))
+	ctx.Logger().Info(fmt.Sprintf("IBC Chan Open cpAppVersion %s: ", cpAppVersion))
+	return version, nil
+}
+
+// OnChanOpenAck implements the IBCModule interface
+func (im IBCModule) OnChanOpenAck(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+	counterpartyChannelId string, // counterpartyChannelId
+	counterpartyVersion string,
+) error {
+	// core/04-channel/types contains a helper function to split middleware and underlying app version
+	// _, _ := channeltypes.SplitChannelVersion(counterpartyVersion)
+	// doCustomLogic()
+	// call the underlying applications OnChanOpenTry callback
+	return im.app.OnChanOpenAck(ctx, portID, channelID, counterpartyChannelId, counterpartyVersion)
+}
+
+// OnChanOpenConfirm implements the IBCModule interface
+func (im IBCModule) OnChanOpenConfirm(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	// doCustomLogic()
+	return im.app.OnChanOpenConfirm(ctx, portID, channelID)
+}
+
+// OnChanCloseInit implements the IBCModule interface
+func (im IBCModule) OnChanCloseInit(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	// doCustomLogic()
+	return im.app.OnChanCloseInit(ctx, portID, channelID)
+}
+
+// OnChanCloseConfirm implements the IBCModule interface
+func (im IBCModule) OnChanCloseConfirm(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	// doCustomLogic()
+	return im.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnRecvPacket implements the IBCModule interface. A successful acknowledgement
+// is returned if the packet data is successfully decoded and the receive application
+// logic returns without error.
+func (im IBCModule) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) ibcexported.Acknowledgement {
+	// NOTE: acknowledgement will be written synchronously during IBC handler execution.
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return channeltypes.NewErrorAcknowledgement(err.Error())
+	}
+
+	// to be utilized from ibc-go v5.1.0
+	if data.Memo == "stakeibc/LiquidStake" {
```

Comment:
It's not clear to me why you look for either the string "stakeibc/LiquidStake" in `data.memo` with an address in `data.Receiver` or for "address|stakeibc/LiquidStake" in `data.Receiver`. Shouldn't one of the other be enough - why do we need to allow for both options? Your comment implies something to do with ibc-go v5.1.0. Could you explain in the comment and docs please?

Reply:
data.Memo field is used since ibc-go v5.1.0 and therefore - this is added for future use, on v3, it's using data.Receiver field

Reply:
Thanks, that makes sense. Could you add this to the readme for the app router module please?